android/app/src/main/java/org/pathcheck/covidsafepaths/exposurenotifications/ExposureNotificationClientWrapper.java

```diff
@@ -57,26 +61,72 @@ public static ExposureNotificationClientWrapper get(Context context) {
     config = new ExposureConfigurations(context);
   }
 
-  public Task<Void> start(ReactContext context) {
-    return exposureNotificationClient.start()
-        .addOnSuccessListener(unused -> onExposureNotificationStateChanged(context, true))
+  public ListenableFuture<Void> requestPermissionToStartTracing(ReactContext context) {
+    return CallbackToFutureAdapter.getFuture(completer -> exposureNotificationClient.start()
+        .addOnSuccessListener(unused -> {
+          if (startTracingCompleter != null) {
+            startTracingCompleter.set(unused);
+            startTracingCompleter = null;
+          }
+
+          onExposureNotificationStateChanged(context, true);
+          completer.set(unused);
+        })
         .addOnFailureListener(exception -> {
+          if (exception instanceof ApiException) {
+            ApiException apiException = (ApiException) exception;
+            if (apiException.getStatusCode() == ExposureNotificationStatusCodes.RESOLUTION_REQUIRED) {
+              startTracingCompleter = completer;
+              showPermissionDialog(context, apiException, RequestCodes.REQUEST_CODE_START_EXPOSURE_NOTIFICATION);
+              return;
+            }
+          }
+
+          if (startTracingCompleter != null) {
+            startTracingCompleter.setException(exception);
+            startTracingCompleter = null;
+          }
+
           onExposureNotificationStateChanged(context, false);
-          if (!(exception instanceof ApiException)) {
-            return;
+          completer.setException(exception);
+        })
+    );
+  }
+
+  /**
+   * Gets recent (initially 14 days) Temporary Exposure Keys from Google Play Services.
+   */
+  public ListenableFuture<List<TemporaryExposureKey>> requestPermissionToGetExposureKeys(ReactContext context) {
```

Comment:
Just curious if this one could follow the same pattern that the `fetchDiagnosisKeys` having an `onSuccess` `onFailure` implementations.

Reply:
We have a mix of implementations because the GAEN API uses `Tasks` and the app uses `Futures`.
In a future refactor I'll change `ExposureNotificationClientWrapper` to only expose `Futures` (since Tasks can only be created by the API).

That's why in some places we have `addOnSuccessListener`/`addOnFailureListener` and some others have `onSuccess`/`onFailure`